lib/functional/functional_gen.nit

```diff
@@ -0,0 +1,166 @@
+        end
+        return args
+end
+
+
+# Base class to writes Nit's functional classes as text.
+class FunTypeWriter
+
+        # Name of the class
+        var name: String
+
+        # Super classes as text
+        var supers: SequenceRead[String]
+
+        # The arity of the class
+        var nb_formal_types: Int
+
+        # Adds another formal type to express a return type
+        var with_return = false
+
+        # Annotation to add after `call`
+        var annotation = "is abstract"
+
+        # Writes a class from provided attributes
+        fun write(writer: WriterLn)
+        do
+                var i = nb_formal_types
+                var generics = gen_generics(i)
+                if with_return then
+                        generics.push("RESULT")
+                end
+                writer.write("interface {name}{i}")
+                writer.writeln("[" + generics.join(",") + "]")
+                writer.tabs += 1
+                for zuper in supers do
+                        writer.writeln("super {zuper}")
+                end
+                writer.tabs -= 1
+                var output = ""
+                if with_return then
+                        output = generics.pop
+                end
+                var params = new Array[String]
+                for g in generics do
+                        params.push(g.to_lower + ": " + g)
+                end
+
+                var signature = params.join(",")
+                if i == 0 then
+                        writer.write("\tfun call")
+                else
+                        writer.write("\tfun call(" + signature + ")")
+                end
+                if with_return then
+                        writer.write(": {output}")
+                end
+                writer.writeln(" {annotation}")
+                writer.writeln("end")
+        end
+
+end
+
+# Writes all functional types
+fun generate_functypes(n: Int, writer: WriterLn)
+do
+        writer.writeln("""
+# This file is part of NIT ( http://www.nitlanguage.org ).
+#
+# Copyright 2019 Louis-Vincent Boudreault <lv.boudreault95@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""")
+        writer.writeln("""
+# This module provides functional type to represents various function forms.
+# Function types can hold up to 20 arguments. The type `Fun` is for function
+# (input and output) and `Proc` is for procedure (input but no output).
+module functional_types
+""")
```

Comment:
What do you think about adding a note that: "This file is generated, do not edit it manually. Use <insert cmd here> to regenerate it"?